### ISSUE TITLE
File indexing fails when Rstudio project is on a network drive

### ISSUE THREAD

**Issue by kendonB** [rstudio/rstudio#2414]
```r
Warning message:
File monitoring failed for project at "X:/network_path"
Error 1 (Incorrect function)
Features disabled: R source file indexing, Diagnostics
```

I am on Windows 7 x64 Rstudio 1.1.442. The network drive is shared from a parent Ubuntu machine. The Windows 7 OS is a virtualbox guest.

**Comment by rmsharp** [comment]
I have the same issue (pseudo network share with host OS) with Windows 10 x64 Rstudio 1.1.447 Virtualbox 5.2.12 r122591 (Qt5.6.3)

**Comment by mik3y64** [comment]
I just got hit with this exact same issue. I am using Windows Subsystem for Linux (WSL) mapped network drive. 

What is the issue? or how do we suppress the message?

```
Warning message:
File monitoring failed for project at "Z:/path/to/network_drive"
Error 1 (Incorrect function)
Features disabled: R source file indexing, Diagnostics
```

**Comment by kevinushey** [comment]
Most likely, it implies that R / RStudio briefly lost access to the network drive (e.g. it was temporarily unmounted / disconnected) and that caused the RStudio file monitor to shut down.

It's worth considering if we could more gracefully recover in these sorts of situations.

**Comment by AlfonsoMuskedunder** [comment]
I'm experiencing the same error. Could it be that errors when knitting-to-pdf a file living on a network drive are related? (See for instance https://github.com/rstudio/rmarkdown/issues/1385#issuecomment-400699370)

**Comment by rappster** [comment]
Experiencing the same error with using a mounted network drive as established by <https://my.pcloud.com/> 

There might be other factors at play that are not in the realm of RStudio, but it seems to me that RStudio causes pCloud to shut down whenever I either run `renv::install()` (which in my case due to (still) fails with an `Error: source file '<pkg>/renv/staging/2/<dependency>' does not exist`) and/or I try calling `devtools::load_all()`.

The network drive connection is otherwise completely stable. Any hints as to what might be causing this. Also tried this on Linux (Pop!_OS 18.10) and saw pCloud crashing as well.

**Comment by kevinushey** [comment]
Looking again, I believe this call is failing:

https://github.com/rstudio/rstudio/blob/39503aeab68f6da717ceadc25a7b9b9478e0fae8/src/cpp/core/system/file_monitor/Win32FileMonitor.cpp#L478

The [documentation](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-readdirectorychangesw) states:

> If the network redirector or the target file system does not support this operation, the function fails with ERROR_INVALID_FUNCTION.

So ultimately you might just need to disable these features (project indexing, diagnostics) when working on network drives like these.

@rappster: can you confirm whether the issue reproduces when using R outside of RStudio; e.g. using R in the terminal or an alternate GUI?

**Comment by rappster** [comment]
@kevinushey 

Thanks for taking the time to address this.

Strangely, I currently can't reproduce the crashes, at least on Linux (Pop!_OS 19.10). But I might very well be "doing the wrong things" for toubleshooting possibly problematic indexing or diagnostics behavior when using network drives. It'd be great if you could point me to some explicit actions/commands to test.

Here's my troubleshooting process:

1. Disabled the indexing of source files:
![image](https://user-images.githubusercontent.com/1205659/71666370-d14fce00-2d60-11ea-97aa-5c83d016aa0d.png)

2. Turned of code diagnostics:
![image](https://user-images.githubusercontent.com/1205659/71666404-f2182380-2d60-11ea-9964-dc0c20fdaf81.png)

3. Within RStudio, ran `renv::init()` and `renv::install("devtools")` on a toy project: 
Worked fine

4. Ran `devtools::load_all()`:
Worked fine

5. Re-enabled code diagnostics, quit and re-opened RStudio, ran `devtools::load_all()` and `renv::update()`: 
Worked fine

6. Re-enabled source file indexing, quit and re-opened RStudio, quit and re-opened RStudio, ran `devtools::load_all()` and `renv::update()`: 
Worked fine

7. Opened a shell, `cd` to my toy project, opened a R terminal and ran `devtools::load_all()` and `renv::update()`: 
Worked fine

I will try on Windows 10 as well and report back how things went there.

**Comment by kevinushey** [comment]
Right, my point is that if you disable project indexing then you should no longer see this warning:

```
Warning message:
File monitoring failed for project at "X:/network_path"
Error 1 (Incorrect function)
Features disabled: R source file indexing, Diagnostics
```

Of course, it does imply features that depend on project indexing will no longer work (e.g. <kbd>Cmd + .</kbd> to go to file)

**Comment by mikebessuille** [comment]
Consider for 1.5.  Part of "cloud-native" perhaps, and a lot of customers have seen this.

**Comment by ellisifnygaard** [comment]
I recently started getting this warning message multiple times a day:  
  
```
Warning message:
File monitoring failed for project at "Z:/network_path/project_directory"
Error 59 (An unexpected network error occurred)
Features disabled: R source file indexing, Diagnostics
```
Between having to disable source file indexing + diagnostics or put up with the irksome warnings, the latter seems to be the lesser of two evils for the time being. 

It would be great if you manage to address this in a coming release! 🙏

**Comment by neellab-umd** [comment]
I got this error this morning upon waking up my computer

Warning message:
File monitoring failed for project at "my project path"
Error 2 (The system cannot find the file specified)
Features disabled: R source file indexing, Diagnostics, Copilot indexing

R version 4.3.1
Rstudio 2023.09.0 Build 463
Windows 10
Files on Google Drive synced with my computer

This seems to be related to the issues with randomly not being able to save files that has come up for me in this version of RStudio.

Is there any hope of getting these issues solved? It is making it hard to work in RStudio, which I do much of every day.

**Comment by jeffvroom** [comment]
For improved resilience, I believe the code should not disable indexing permanently, but instead should periodically try this operation again, at least for some time period during startup. It sounds like in some cases at least, this is an intermittent problem.

**Comment by llrs** [comment]
I received a similar error message in Ubuntu in a folder that is not on a network drive.
It is not reproducible, as this is the first time I see this in this project:

```
File monitoring failed for project at "~/ShinyApps/EpiNeuroCovid"
Error 28 (inotify_add_watch: No watches available)
Features disabled: R source file indexing, Connect, Diagnostics
```

R 4.4.0
RStudio 2024.04.1+748 "Chocolate Cosmos" Release (3ada7c6ddc8fcdb86a727a4f0ae467b9d9a7296c, 2024-05-07) for Ubuntu Jammy
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) rstudio/2024.04.1+748 Chrome/120.0.6099.291 Electron/28.2.6 Safari/537.36, Quarto 1.4.346 (/opt/quarto/bin/quarto)

The message is similar to an [old post](https://forum.posit.co/t/rstudio-file-monitoring-failure/52272). I am not sure how to interpret the suggested output:

```
  INOTIFY
   WATCHER
    COUNT     PID     CMD
----------------------------------------
   41977   299050  /usr/lib/rstudio/resources/app/bin/rsession --config-file none --program-mode desktop --www-port 39851 --launcher-token 
   41977   299049  /usr/lib/rstudio/resources/app/bin/rsession --config-file none --program-mode desktop --www-port 39851 --launcher-token 
   41977   299048  /usr/lib/rstudio/resources/app/bin/rsession --config-file none --program-mode desktop --www-port 39851 --launcher-token 
   20632   318765  /usr/lib/rstudio/resources/app/bin/rsession --config-file none --program-mode desktop --www-port 15569 --launcher-token 
```

**Comment by kevinushey** [comment]
Are you working within a project with a very large number of files? It is surprising that the inotify watcher count is being exhausted like this.

**Comment by llrs** [comment]
@kevinushey No, the project at ~/ShinyApps/EpiNeuroCovid has just 9 files. But I have another project (that could have been open at the same time) which currently has 95282 files most of them (67580) in a folder in .gitignore (the output of other software).

**Comment by ngctnnnn** [comment]
I received the same notification when conducting reading and writing continuously on R, especially when I run multi-processing via `plan(multisession)`
```bash
Warning message:
File monitoring failed for project at "R:/path_to_NAS_system"
Error 2 (The system cannot find the file specified)
Features disabled: R source file indexing, Connect, Diagnostics
```
